### ISSUE TITLE
Should we include typescript in peerDependencies?

### ISSUE THREAD

**Issue by davidmytton** [arcjet/arcjet-js#1029]
I had a user report a compile error which turned out to be because they were using TypeScript 4.9.5. Updating to the latest version (5.5.2) solved this. Should we add TypeScript >= to peerDependencies?

```
21:28:15.573 • Next.js 14.2.3
--
21:28:15.574 |  
21:28:15.575 | Linting and checking validity of types
21:28:24.363 | Failed to compile.
21:28:24.363
21:28:24.364 /node_modules/-pnpm/@arcjet+next@1.0.0-alpha.13_@bufbuild+protobuf@1.10.0_@connectrpc+connect@1.4.0_next@14.2.3/node_modules/@arcjet/next/index.d.ts:84:36
21:28:24.364 Type error: Type parameter declaration expected.
21:28:24.364 |  
21:28:24.364 | 82 | * @param options - Arcjet configuration options to apply to all requests.
21:28:24.364 | 83 | * /
21:28:24.364 | > 84 export default function arcjetNext<const Rules extends (Primitive | Product)[]>(options: ArcjetOptions<Rules>): ArcjetNext<Simpify<ExtraProps<Rules>>>;
21:28:24.364 |  
21:28:24.364 | 85 /**
21:28:24.364 | 86 | * Protects your Next.js application using Arcjet middleware. 
21:28:24.364 | 87
21:28:24.399 | ELIFECYCLE Command failed with exit code 1.
21:28:24.420 | Error: Command "pnpm run build" exited with 1
21:28:24.977

```

**Comment by blaine-arcjet** [comment]
I don't believe I've ever seen a project specify typescript as a peerDependency so I don't think we should.

Additionally, peerDependencies wouldn't solve the issue here because the newer TypeScript would be installed in `node_modules/@arcjet/next/node_modules/typescript` and they would still be compiling with typescript 4.9.5 which they installed at the top level.

**Comment by wooorm-arcjet** [comment]
I second Blaine.

TS is more like Node or Bun or another engine. And a proper solution then is, I believe, to have:

a) some policy on how far back to support TS, e.g., ±2 years (sadly there is nothing like [`nodejs/release`](https://github.com/nodejs/release#nodejs-release-working-group) for TS),
b) document that somewhere,
c) somehow test that policy (far future nice to have)
d) only update the supported version in major releases

In my OSS work I occasionally run into users using ancient TS. But I think almost all folks figure this out themselves. So we don’t have an explicit policy/docs/tests for this.

But we do for Node. For example: https://github.com/mdx-js/mdx/tree/main/packages/rollup#compatibility. And then testing are run in GH actions across both outliers (oldest, latest).

**Comment by davidmytton** [comment]
Is there a common timeline for commercial projects to maintain support for older TS versions? Perhaps we can align with the TS version that was released at the time of the current Node LTS?

**Comment by wooorm-arcjet** [comment]
Not that I am aware of!

Did some digging into both release processes, here’s my findings:

* For TS, it looks like there’s a “minor” (e.g., 5.7, 5.8) about every 3 months
* Semver doesn’t mean much to TS: after 4.9 they cut 5.0, with maybe more breaking changes/features, but every minor also introduces more/less type errors, so every release is breaking anyway

That seems to imply that a “major” release line (e.g., 4, 5) lasts 2.5 years.

Interestingly, an even (e.g., 22, 24) release of Node is the same. 12 months active, 18 months maintenance mode.

One more thing to think about is that an idea of “Node 18 is supported” or so, does not always mean that `18.0.0` exactly is supported. Presumably there was an important bug fix in `18.0.1` 2 days later or so which everyone should have. And, probably, there were also some important backported features on later minors.

And: TS is a dependency, specifically a developer tool. I can see why Node.js would need a wider range of supported versions, whereas with `typescript` the range can be more narrow.

What I would suggest, is when bumping a major, taking the lowest currently maintained Node, so today that would be Node 20.
And then taking for example the current TS version minus 4 minor releases, or so? So today that would be 5.8 minus 4 is 5.4?

If with Node 20 today (active maintenance), taking the initial release date of 2023-04-18, we’d get [TS 5.0](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/).

If with Node 22 today (lts), taking the initial release date of 2024-04-24, we’d get [TS 5.4](https://devblogs.microsoft.com/typescript/announcing-typescript-5-4/).

References:

* https://github.com/microsoft/TypeScript/wiki/TypeScript's-Release-Process#how-often-does-typescript-release
* https://github.com/nodejs/release#release-plan

**Comment by davidmytton** [comment]
I'd suggest we go with TS 5 and keep up with the latest minor releases until the next breaking change. When TS 6 comes out, we can decide if we should upgrade or not based on the release notes.

**Comment by blaine-arcjet** [comment]
> keep up with the latest minor releases until the next breaking change

TypeScript releases breaking changes in each minor release. The only place they *shouldn't* is patch releases, but I'm not sure they hold themselves to that 100% of the time.

**Comment by blaine-arcjet** [comment]
I think we should always be using the latest version of TypeScript. It's generally easy to upgrade between minor versions, but it can start to be difficult once you start to fall behind (such as the 4.x -> 5.x update that inspired this issue).

**Comment by davidmytton** [comment]
> I don't believe I've ever seen a project specify typescript as a peerDependency so I don't think we should.

So we have an informal policy that we're always on latest then? And don't include it in peerdeps?

**Comment by blaine-arcjet** [comment]
> So we have an informal policy that we're always on latest then? And don't include it in peerdeps?

Yeah, I think that's the best way forward. We could opt for only using "new" features introduced in the major bump but that might limit us if we need a feature introduced in a minor bump.

It is also worth noting that while the root comment flagged a feature in `.d.ts` that was unsupported in TS 4.x, this was a bigger issue when we shipped `.ts` files in the published packages, which we no longer include.

**Comment by wooorm-arcjet** [comment]
Can this then be closed? If not yet, should this be documented somewhere?

**Comment by blaine-arcjet** [comment]
I believe we want to document the support level and close this with that PR.

**Comment by wooorm-arcjet** [comment]
Made a PR to document it in the main readme, if that’s what you meant. Though an ADR is also possible I guess?
